src/Analysis/Codelyzer.Analysis.Build/ExternalReferenceLoader.cs

```diff
@@ -0,0 +1,215 @@
+                        {
+                            externalReference.Identity = symbol.Identity.Name;
+                            externalReference.Version = symbol.Identity.Version != null ? symbol.Identity.Version.ToString() : string.Empty;
+                            name = symbol.Identity.Name;
+                        }
+
+                        var nugetRef = _externalReferences.NugetReferences.FirstOrDefault(n => n.Identity == name);
+
+                        if (nugetRef == null)
+                        {
+                            nugetRef = _externalReferences.NugetReferences.FirstOrDefault(n => filePath.ToLower().Contains(string.Concat(Constants.PackagesDirectoryIdentifier, n.Identity.ToLower(), ".", n.Version)));
+                        }
+
+                        if (nugetRef != null)
+                        {
+                            //Nuget with more than one dll?
+                            nugetRef.AssemblyLocation = filePath;
+
+                            //If version isn't resolved, get from external reference
+                            if (string.IsNullOrEmpty(nugetRef.Version) || !Regex.IsMatch(nugetRef.Version, @"([0-9])+(\.)([0-9])+(\.)([0-9])+"))
+                            {
+                                nugetRef.Version = externalReference.Version;
+                            }
+                        }
+                        else if (filePath.Contains(Common.Constants.PackagesDirectoryIdentifier, System.StringComparison.CurrentCultureIgnoreCase))
+                        {
+                            _externalReferences.NugetDependencies.Add(externalReference);
+                        }
+                        else if (!projectReferenceNames.Any(n => n.StartsWith(name)))
+                        {
+                            _externalReferences.SdkReferences.Add(externalReference);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.LogError(ex, "Error while resolving reference {0}", externalReferenceMetaData);
+                    }
+                }
+            }
+        }
+        private bool LoadFromNugetFile(string filePath)
+        {
+            var foundPackage = false;
+            var packageDirSplit = filePath.Split(new string[] { Constants.PackagesDirectoryIdentifier }, StringSplitOptions.RemoveEmptyEntries);
+            
+            if(packageDirSplit.Length > 1)
+            {
+                var packageFolderName = packageDirSplit[1].Split(Path.DirectorySeparatorChar)[0];
+                var packageFolderDir = Path.Combine(packageDirSplit[0], Constants.PackagesFolder, packageFolderName);
+                var nupkg = Directory.EnumerateFiles(packageFolderDir, Constants.NupkgFileExtension, SearchOption.AllDirectories).ToList();
+                nupkg.ForEach(file => {
+                    var reader = new PackageArchiveReader(file);
```

Comment:
We are handling it on the individual package level (the function calling this). Any exception in this will cause a single nuget parsing to fail and continue with the remaining ones. I removed the foreach here because a nuget will have one nupkg file only (the foreach was misleading).